### ISSUE TITLE
Frame latency in WebGPU

### ISSUE THREAD

**Issue by floooh** [gpuweb/gpuweb#556]
@kangz suggested that I write an issue here:

I haven't thought this through completely, so please excuse any unfocused rumblings :)

TL;DR: I think WebGPU implementations should guarantee a fixed frame latency (so it doesn't change frame latency once an application is running), specify a 'worst case latency' in the spec which implementations must adhere too, and (maybe) provide a way to as an implementation at startup for the frame latency. Plus (of course), implementations should strive to keep the frame latency as low as possible.

Why:

Apart from obvious user-experience aspects (low latency == better), e.g. button-to-screen, or head-movement in AR/VR, frame latency also affects the coding side now because WebGPU offers more explicit resource management than WebGL.

Consider dynamic resource updates via async-mapped buffers allocated on demand:

- at the start of a frame, app code if no mapped buffer is available, a new buffer is created as mapped
- ...during the frame, the buffer is filled with new data
- before queue-submit, the buffer is unmapped, and optionally a copy/blit-operation is recorded in the command encoder
- after queue-submit, a new async-mapping is requested immediately

With a fixed frame-latency, this creates a new buffer each frame at the start of the application, the number depending on the frame latency of the WebGPU implementation.

If the frame-latency stays fixed during the entire lifetime of the application, no buffers will be created except during the first N frames.

Now to the problems if the frame latencies varies between WebGPU implementations or even during the lifetime of an application:

- Memory usage basically becomes unpredictable, the total size of staging buffers can be quite big, and each additional frame latency adds a non-trivial amount of memory usage (not in the WASM heap, but still...), and most of this memory just sits around waiting in the frame pipeline until it's their turn... the longer the 'frame pipeline' the worse.
- If the frame-latency would be completely unpredictable, internal data structures in the app code must be dynamically sized (e.g. queues or ring buffers holding references to the staging buffers).

So ideally we'd have:

- a worst-case frame latency in the spec, so that applications can estimate memory usage and use this compile-time size of data structures
- a runtime-frame-latency by a given implementation in a given environment that can be queried at startup and doesn't change (which applications can use to tweak their behaviour)
- a guarantee that frame latency is either fixed throughout the application lifetime, or at least doesn't get bigger than the 'worst case latency' across all the implementations and the runtime-latency of a given implementation...

As I said above, there may be other implications/advantages/disadvantages around frame latency that I haven't thought of, but I think it's at least important to raise the issue so it can be discussed :)

**Comment by floooh** [comment]
PS afterthought: naturally, the role of GPUFences should be considered too... As far as I understand it, fences can be used to prevent the application code from running too far ahead of the GPU (e.g. stop submitting new work or demand-allocate new staging buffers if the CPU is too far ahread), but they can't help with the "built-in" minimal frame latency of a WebGPU implementation (e.g. if a WebGPU implementation has 2 frame latency, there's no way around allocating 3 round-robin buffers, 2 in-flight and one currently written by the CPU).

E.g. if the frame latency is fixed anyway and the CPU side is throttled by requestAnimationFrame() (e.g. vsync), fences are not all that useful(?)

**Comment by kvark** [comment]
@floooh thank you for filing this!

So, if I understand you correctly, the downside you are worried about is the memory consumption? We are looking at, basically, difference between 2x something and 3x something (related to staging) in most cases. Does it warrant an API feature?

> internal data structures in the app code must be dynamically sized (e.g. queues or ring buffers holding references to the staging buffers).

In the way `mapWriteAsync` is proposed today, you'd likely want that number of staging buffers growing dynamically anyway ("the staging belt in user space"). Could you provide more details about the case where you'd strictly want a fixed number of staging buffers?

**Comment by floooh** [comment]
@kvark yeah, I'm mainly concerned about the upper bound for memory consumption, e.g. 2x vs 3x is ok, but it shouldn't be allowed to be higher. There might also be an issue for use in VR/AR if the frame latency is too high (I'm not an expert in that though so I can't really provide useful input).

As long as there are no edge cases where the frame latency can "jump" to a higher value I think that's ok (e.g. something "gets stuck", maybe when putting a tab into background, so that I'm "blindly" submitting new frames, and creating new staging buffers but the backend cannot consume them).

IMHO ideally in a "typical" rendering scenario, e.g. rendering at vsync interval inside requestAnimationFrame(), no special syncing (e.g. via fences) should be required by the WebGPU programmer to prevent any such spikes (if they can happen).

**Comment by litherum** [comment]
It sounds like you’re advocating for the *existence* of a latency-consistent codepath. It doesn’t sound like you’re advocating for every single WebGPU feature to have a fixed latency.

For example, I believe an application which uses mapWriteAsync() would have fixed latency, but an application which uses setSubData() might not (there may be internal allocations). This still passes your test, right?

Or, perhaps even setSubData() passes your test because the number of internal allocations is bounded?

**Comment by litherum** [comment]
I guess another way to ask my question:

Can you be precise about what, specifically, you would like the CG to do?

**Comment by Kangz** [comment]
I think @floooh would like the CG to fix an upper bound of how many frames can be in flight at the same time. Basically if you signal a fence with the frame number at the end of each rAF, what's the maximum difference between the current frame number and `fence.getCompletedValue()`. Alternatively the CG can decide there is no such guarantee.

**Comment by litherum** [comment]
Okay, that helps.

Putting a specific number in the spec would be the wrong thing to do. The number of backbuffers in the browser compositor’s frame buffer is an implementation detail. Browsers should be free to use single buffering if they want, double buffering if they want, triple buffering if they want, and septuple buffering if they want. Any specific number we could write in the spec would be wrong. It also needs to be able to change throughout the lifetime of the page, if you, for example, drag the browser window from one monitor to another one.

The concept of frame numbers isn’t really relevant in a compute-only workflow. I’m not sure how you could make any such guarantee there.

Also, given that shaders might describe arbitrary computation, they may take arbitrary time to execute. Including longer than a few frames. I don’t think we can place any limit on the complexity of the shader because 1) that’s incomputable in the general case and 2) Different devices have different performances, so we can’t statically know whether some computation will last longer than a frame on a given device without actually executing it. The timeline of when the browser/OS will kill long-running shaders is an implementation detail, which different platforms may intentionally choose to set differently. It shouldn’t be in a number in the spec.

Rather than putting a specific number in the spec, I am willing to say that the number, if it were to exist, would be _bounded_. Sort of like the difference between LR(3) grammars and LR(k) grammars. I think that’s already true of every implementation, so it would just be making the spec reflect reality, and giving authors some assurances that the browser will do something sensible with their content.

**Comment by floooh** [comment]
Yeah, the main reason why I think this is even noteworthy is that I was a bit surprised that the current Dawn implementation seems to have more than one frame in-flight. In WebGL, and with "convenience" functions like buffer.setSubData() I don't have a choice than trusting the underlying implementation to not do anything stupid, and the frame latency WebGL isn't even observable from the outside anyway (at least not trivially).

With map*Async() and GPU-fences, the frame latency becomes visible to the API user, so IMHO the WebGPU spec shouldn't ignore it (even a sentence that WebGPU implementations do not provide any guarantees is better than not mentioning it).

Maybe the spec is the wrong place for anything beyond this, maybe some sort of "implementation notes" or "implementation addendum" is the better place, but it would be good to have a place with 'guidelines', 'notes' or 'rules' about expected behaviour.

My main concern is basically that an API user write "dumb^H^H^Hstraightfoward" on-demand buffer-allocation code which allocates 2 or 3 buffer under normal circumstances, but then there's some unexpected edge case where (for whatever reason), the CPU-side runs ahead and suddenly allocated dozens of buffers.

(PS: my comment overlapped with your last one @litherum, and didn't take your comment into account)

**Comment by litherum** [comment]
@floooh i think we’re in agreement.

**Comment by litherum** [comment]
(Having more than one frame in flight at a time is the common case on triple-buffered systems.)
